kupo-nixos-module.nix

```diff
@@ -0,0 +1,126 @@
+# NixOS module for kupo service
+{ config, lib, pkgs, ...}:
+let
+  cfg = config.services.kupo;
+in  with lib; {
+  options.services.kupo = with types; {
+    enable = mkEnableOption "Kupo - fast, lightweight and configurable chain-index";
+
+    package = mkOption {
+      description = "Kupo package.";
+      type = package;
+    };
+
+    user = mkOption {
+      description = "User to run kupo service as.";
+      type = types.str;
+      default = "kupo";
+    };
+
+    group = mkOption {
+      description = "Group to run kupo service as.";
+      type = types.str;
+      default = "kupo";
+    };
+
+    nodeSocket = mkOption {
+      description = "Path to cardano-node IPC socket.";
+      type = str;
+    };
+
+    nodeConfig = mkOption {
+      description = "Path to cardano-node config.json file.";
+      type = str;
+    };
+
+    host = mkOption {
+      description = "Host address or name to listen on.";
+      type = str;
+      default = "localhost";
+    };
+
+    extraArgs = mkOption {
+      description = "Extra arguments to kupo command.";
+      type = listOf str;
+      default = [ ];
+    };
+  };
+
+  config = mkIf cfg.enable {
+    assertions = mkIf (config ? services.cardano-node && config.services.cardano-node.enable) [
+      {
+        assertion = config.services.cardano-node.systemdSocketActivation;
+        message = "The option services.cardano-node.systemdSocketActivation needs to be enabled to use Kupo with the cardano-node configured by that module.";
+      }
+    ];
+
+    # get configuration from cardano-node module if there is one
+    services.kupo = mkIf (config ? services.cardano-node && config.services.cardano-node.enable) {
+      nodeSocket = mkDefault config.services.cardano-node.socketPath;
+      # hacky way to get cardano-node config path from service
+      nodeConfig = mkDefault (builtins.head (
+        builtins.match ''.* (/nix/store/[a-zA-Z0-9]+-config-0-0\.json) .*''
+          (builtins.readFile (builtins.replaceStrings [ " " ] [ "" ] config.systemd.services.cardano-node.serviceConfig.ExecStart))
+      ));
+    };
+
+    users.users.kupo = mkIf (cfg.user == "kupo") {
+      isSystemUser = true;
+      group = cfg.group;
+      extraGroups = [ "cardano-node" ];
+    };
+    users.groups.kupo = mkIf (cfg.group == "kupo") { };
+
+    systemd.services.kupo = {
+      enable = true;
+      after = [ "cardano-node.service" ];
+      wantedBy = [ "multi-user.target" ];
+
+      script = escapeShellArgs (concatLists [
+        [ "${cfg.package}/bin/kupo" ]
+        [ "--node-socket" cfg.nodeSocket ]
+        [ "--node-config" cfg.nodeConfig ]
+        [ "--host" cfg.host ]
+        # [ "--workdir" "/tmp" ]
+        [ "--in-memory" ]
+        [ "--match" "*/*" ]
+        [ "--since" "origin" ]
+        cfg.extraArgs
+      ]);
+
+      serviceConfig = {
+        User = cfg.user;
+        Group = cfg.group;
+        # Security
```

Comment:
please run `systemd-analyze security` and try to improve based on suggestions there, unless the service breaks.